src/libopenrave/robot.cpp

```diff
@@ -559,42 +559,81 @@ void RobotBase::RobotStateSaver::Release()
     KinBodyStateSaver::Release();
 }
 
-///\brief removes the robot from the environment temporarily while in scope
-class EnvironmentBodyRemover
-{
-public:
-    EnvironmentBodyRemover(OpenRAVE::KinBodyPtr pBody) : _pBody(pBody), _grabbedStateSaver(pBody, OpenRAVE::KinBody::Save_GrabbedBodies) {
-        if( _pBody->IsRobot() ) {
-            // If the manip comes from a connected body, the information of which manip is active is lost once the robot
-            // is removed from env. Need to save the active manip name so that we can set it back later when the robot
-            // is re-added to the env.
-            _pBodyRobot = OpenRAVE::RaveInterfaceCast<OpenRAVE::RobotBase>(_pBody);
-            if( !!_pBodyRobot->GetActiveManipulator() ) {
-                _activeManipName = _pBodyRobot->GetActiveManipulator()->GetName();
-            }
+EnvironmentBodyRemover::EnvironmentBodyRemover(KinBodyPtr pBody, int restoreOptions) : _pBody(pBody), _environmentBodyIndex(_pBody->GetEnvironmentBodyIndex()), _restoreOptions(restoreOptions) {
+    if( _pBody->IsRobot() ) {
+        // If the manip comes from a connected body, the information of which manip is active is lost once the robot
+        // is removed from env. Need to save the active manip name so that we can set it back later when the robot
+        // is re-added to the env.
+        _pBodyRobot = RaveInterfaceCast<RobotBase>(_pBody);
+        if( !!_pBodyRobot->GetActiveManipulator() ) {
+            _activeManipName = _pBodyRobot->GetActiveManipulator()->GetName();
         }
-        _pBody->GetEnv()->Remove(_pBody);
     }
+    _pBody->GetGrabbedInfo(_pGrabbedInfos);
+    _pBody->GetEnv()->Remove(_pBody);
+}
 
-    ~EnvironmentBodyRemover() {
-        _pBody->GetEnv()->Add(_pBody, IAM_StrictNameChecking);
-        _grabbedStateSaver.Restore();
-        _grabbedStateSaver.Release();
-        if( !!_pBodyRobot && !_activeManipName.empty() ) {
-            OpenRAVE::RobotBase::ManipulatorPtr pmanip = _pBodyRobot->GetManipulator(_activeManipName);
-            // it might be ok with manipulator doesn't exist if ConnectedBody acitve state changes.
-            if( !!pmanip ) {
-                _pBodyRobot->SetActiveManipulator(pmanip);
+EnvironmentBodyRemover::~EnvironmentBodyRemover() noexcept(true) {
+    _pBody->GetEnv()->AddKinBody(_pBody, IAM_StrictNameChecking, _environmentBodyIndex);
```

Comment:
@eisoku9618 @kanbouchou 

Just to confirm.
Is it ok to always set `_environmentBodyIndex` when adding?

Reply:
NVM, this is already answered in other thread.